File: Power_Analysis/Multi_sim.R

library(tidyverse) #for everything 
library(gridExtra) #to export the stats tables
library(grid)
library(cowplot) #to arrange ggplots in a grid 


# Simulate data -----------------------------------------------------------


rm(list = ls()) #clear workspace
set.seed(123)

nsim <- 1000 #number of simulations
MSMRMods <- list() #parking lots for models 
CortMods <- list()
CortMSMR <- list()

#MSMR determined from Goetz 2018 figure 3 control 
#female average 0.170281, standard error 0.0107
#Male average 0.213707571809, standard error 0.0169689
MSMRSD <- 0.015*sqrt(25) #standard deviation = standard error * sqrt(n)

#Cort release rate from figure 4B Gabor 2013 
#SD is estimated by 3/4 x IQR
CortSD <- 2.85
pb <- progress_bar$new(total = length(nsim)) #for progress bar

##for gamma, from lab 5 
#scale parameter = variance/mean = coef.var. 
#shape parameter = mean^2/variance

N <- 72 #sample size
SimulatedData <- data.frame(ID = as.character(seq(1,N,1)), 
                            Temp = as.numeric(rep(c(14,18,22,26,30,34), length = N)), # six temp treatments 
                            sex = NA,
                            Mass = NA) 

#Mass estimates by sex from Goetz 2018 Table 1
for (i in 1:nsim){
  SimulatedData$sex = as.character(sample(c("M","F"), N, replace = TRUE)) #assume equal numbers
  for (j in 1:N) { 
    if (SimulatedData$sex[j] == "M") {
      SimulatedData$Mass[j] <- rgamma(1, shape = 6.8^2/0.4^2, scale = 0.4^2/6.8)
    } else {
      SimulatedData$Mass[j] <- rgamma(1, shape = 20^2/2^2, scale = 2^2/20)
    }
  }
  MSMR_det <- SimulatedData$Mass^-0.25 * exp(SimulatedData$Temp/10) * (0.2/6.04) #6.04 is the average MSMR before it is scaled to Goetz 2013
  Cort_det <- SimulatedData$Mass^-0.25 * exp(SimulatedData$Temp/10) * (7.8/6.04) #7.8 is the average MSMR before it is scaled to Gabor 2018 
  SimulatedData$MSMR <- rgamma(N, shape = (MSMR_det^2)/(MSMRSD^2), scale = (MSMRSD^2)/MSMR_det) 
  SimulatedData$Cort <- rgamma(N, shape = (Cort_det^2)/(CortSD^2), scale = (CortSD^2)/Cort_det) 
  MSMRModel <- lm(log(MSMR) ~ log(Mass) + Temp, data = SimulatedData)
  CortModel <- lm(log(Cort) ~ log(Mass) + Temp, data = SimulatedData)
  CortMSMRModel <- lm(log(Cort) ~ log(MSMR), data = SimulatedData)
  MSMRMods[[i]] <- MSMRModel
  CortMods[[i]] <- CortModel
  CortMSMR[[i]] <- CortMSMRModel
}

#store the final example to visualize one dataframe 
write.csv(SimulatedData, "Power_Analysis/SimulatedData.csv", row.names = FALSE)


# Stats table -------------------------------------------------------------


#blank dataframes for averages
MSMRaverage <- matrix(NA, nrow = 3, ncol = 4)
Cortaverage <- matrix(NA, nrow = 3, ncol = 4)
CortMSMRaverage <- matrix(NA, nrow = 2, ncol = 4)

#get averages across the 1000 simulated models
for (i in 1:3){
  for (j in 1:4){
    MSMRaverage[i, j] <- mean(sapply(1:nsim, function(n) coef(summary(MSMRMods[[n]]))[i, j]))
    MSMRrsq <- mean(sapply(1:nsim, function(n) summary(MSMRMods[[n]])$r.squared))
    Cortaverage[i, j] <- mean(sapply(1:nsim, function(n) coef(summary(CortMods[[n]]))[i, j]))
    Cortrsq <- mean(sapply(1:nsim, function(n) summary(CortMods[[n]])$r.squared))
  }
}

#confidence intervals 
MSMRConfint <- matrix(NA, nrow = 3, ncol = 2)
CortConfint <- matrix(NA, nrow = 3, ncol = 2)

for (i in 1:3){
  for (j in 1:2) {
    MSMRConfint[i, j] <- mean(sapply(1:nsim, function(n) confint(MSMRMods[[n]])[i, j]))
    CortConfint[i, j] <- mean(sapply(1:nsim, function(n) confint(CortMods[[n]])[i, j]))
  }
}

#Cort ~ MSMR model averages
for (i in 1:2){
  for (j in 1:4){
    CortMSMRaverage[i, j] <- mean(sapply(1:nsim, function(n) coef(summary(CortMSMR[[n]]))[i, j]))
        CortMSMRrsq <- mean(sapply(1:nsim, function(n) summary(CortMSMR[[n]])$r.squared))
  }
}

CortMSMRConfint <- matrix(NA, nrow = 2, ncol = 2)

for (i in 1:2){
  for (j in 1:2) {
    CortMSMRConfint[i, j] <- mean(sapply(1:nsim, function(n) confint(CortMSMR[[n]])[i, j]))
  }
}

#organize statistics in a way I like 
StatTab <- rbind(MSMRaverage, Cortaverage, CortMSMRaverage) %>%
  as.data.frame(.) %>%
  slice(-c(1,4,7)) %>% #remove intercept rows
  cbind(., c(MSMRConfint[2,1],
             MSMRConfint[3,1],
             CortConfint[2,1],
             CortConfint[3,1],
             CortMSMRConfint[2,1]), 
        c(MSMRConfint[2,2],
          MSMRConfint[3,2],
          CortConfint[2,2],
          CortConfint[3,2],
          CortMSMRConfint[2,2]), #add back in confidence intervals as column
        c(MSMRaverage[1,1],
          MSMRaverage[1,1],
          Cortaverage[1,1],
          Cortaverage[1,1],
          CortMSMRaverage[1,1]), #add back in intercepts as column
        c(MSMRrsq,
          MSMRrsq,
          Cortrsq,
          Cortrsq,
          CortMSMRrsq)) %>% #and r squared values 
  `colnames<-`(c("Estimate", "SE Est.", "T value",  "p value", "Lower 95% CI", "Upper 95% CI", "Intercept", "R squared")) %>%
  `rownames<-`(c("MSMR ~ mass",
                 "MSMR ~ Temp",
                 "Cort ~ mass", 
                 "Cort ~ Temp",
                 "Cort ~ MSMR")) %>%
  mutate(Intercept = as.numeric(Intercept),
         `R squared` = as.numeric(`R squared`)) %>% #make them numeric for rounding 
  mutate(Estimate = round(Estimate, 3),
         `SE Est.` = round(`SE Est.`, 3),
         `T value` = round(`T value`, 2),
         `p value` = round(`p value`, 3),
         Intercept = round(Intercept, 2),
         `R squared` = round(`R squared`, 3)) %>%
  mutate(`p value` = ifelse(`p value` < 0.001, "< 0.001", `p value`)) #change very small p values to < 0.001
  
  
StatTab$Intercept[c(2,4)] <- " " #repeated values 
StatTab$'R squared'[c(2,4)] <- " "
  

#set the theme for the table png
tt1 <- ttheme_minimal(rowhead=list(fg_params=list(fontface = "bold")))

#export stats table as a csv
write.csv(StatTab, "Power_Analysis/StatsTab.csv", row.names = TRUE)

#export stats table asa pretty png 
png("Power_Analysis/StatsTab.png", 
    height = 190*nrow(StatTab), 
    width = 430*ncol(StatTab),
    res = 300)
grid.newpage()
grid.table(StatTab, theme = tt1)
dev.off()  


# Plots -------------------------------------------------------------------

### BE WARNED ###
#this is not computationally efficient, but it works
#visualizing with the last plot_grid will take a while

#Start with an emptly plot to put expected lines on 
MSMRMassPlot <- ggplot() +
  geom_point(aes(x = log(SimulatedData$Mass), y = log(SimulatedData$MSMR)), color = NA) + #just need this to set the scale for the axis 
  labs(x = "ln Body Mass (g)",
       y = "ln MSMR") +
  scale_y_continuous(limits = c(-7, 0)) +
  theme_classic () 

#plot the 1000 lines for each model
i <- 1
while (i <= length(MSMRMods)){
  MSMRMassPlot <- MSMRMassPlot +
    geom_abline(intercept = summary(MSMRMods[[i]])$coefficients[1], 
                slope = summary(MSMRMods[[i]])$coefficients[2], #2nd row is mass
                color = "blue",
                alpha = 0.02) #fairly clear so we can see overlap
  i <- i + 1
}

MSMRTempPlot <- ggplot() +
  geom_point(aes(x = SimulatedData$Temp, y = log(SimulatedData$MSMR)), color = NA) + #just need this to set the scale for the axis 
  labs(x = "Temperature (C)",
       y = "ln MSMR") +
  theme_classic () 

i <- 1
while (i <= length(MSMRMods)){
  MSMRTempPlot <- MSMRTempPlot +
    geom_abline(intercept = summary(MSMRMods[[i]])$coefficients[1], 
                slope = summary(MSMRMods[[i]])$coefficients[3], #3rd row is temp
                color = "blue",
                alpha = 0.02) #fairly clear so we can see overlap
  i <- i + 1
}

CortMassPlot <- ggplot() +
  geom_point(aes(x = log(SimulatedData$Mass), y = log(SimulatedData$Cort)), color = NA) + #just need this to set the scale for the axis 
  labs(x = "ln Body Mass (g)",
       y = "ln Cort") +
  scale_y_continuous(limits = c(-5, 2)) +
  theme_classic ()

i <- 1
while (i <= length(CortMods)){
  CortMassPlot <- CortMassPlot +
    geom_abline(intercept = summary(CortMods[[i]])$coefficients[1], 
                slope = summary(CortMods[[i]])$coefficients[2], #2nd row is mass
                color = "blue",
                alpha = 0.02) #fairly clear so we can see overlap
  i <- i + 1
}

CortTempPlot <- ggplot() +
  geom_point(aes(x = SimulatedData$Temp, y = log(SimulatedData$Cort)), color = NA) + #just need this to set the scale for the axis 
  labs(x = "Temperature (C)",
       y = "ln Cort") +
  theme_classic ()

i <- 1
while (i <= length(CortMods)){
  CortTempPlot <- CortTempPlot +
    geom_abline(intercept = summary(CortMods[[i]])$coefficients[1], 
                slope = summary(CortMods[[i]])$coefficients[3], #3rd row is temp
                color = "blue",
                alpha = 0.02) #fairly clear so we can see overlap
  i <- i + 1
}

CortMSMRPlot <- ggplot() +
  geom_point(aes(x = log(SimulatedData$MSMR), y = log(SimulatedData$Cort)), color = NA) + #just need this to set the scale for the axis 
  labs(x = "ln MSMR",
       y = "ln Cort") +
  theme_classic ()

i <- 1
while (i <= length(CortMSMR)){
  CortMSMRPlot <- CortMSMRPlot +
    geom_abline(intercept = summary(CortMSMR[[i]])$coefficients[1], 
                slope = summary(CortMSMR[[i]])$coefficients[2], #2nd row is msmr
                color = "blue",
                alpha = 0.02) #fairly clear so we can see overlap
  i <- i + 1
}

#put all plots in a grid 
plot_grid(MSMRMassPlot, MSMRTempPlot, CortMassPlot, CortTempPlot, CortMSMRPlot, 
          labels = c('A', 'B', 'C', 'D', 'E'), label_size = 12)

File: Power_Analysis/Multi_Sim.py
# -*- coding: utf-8 -*-
"""
Created on Tue May 20 23:16:26 2025

@author: kphud
"""

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from tqdm import tqdm
from statsmodels.formula.api import ols
from statsmodels.stats.outliers_influence import summary_table
import os

np.random.seed(123)

# Setup
nsim = 1000
N = 72
temps = np.tile([14, 18, 22, 26, 30, 34], N // 6)
ids = [str(i+1) for i in range(N)]

# Constants
MSMRSD = 0.015 * np.sqrt(25)
CortSD = 2.85

# Storage
MSMR_models = []
Cort_models = []
Cort_MSMR_models = []

# Simulation loop
for _ in tqdm(range(nsim), desc="Simulations"):
    sex = np.random.choice(["M", "F"], size=N)
    mass = np.array([
        np.random.gamma(shape=(6.8**2)/(0.4**2), scale=(0.4**2)/6.8) if s == 'M' else
        np.random.gamma(shape=(20**2)/(2**2), scale=(2**2)/20) for s in sex
    ])
    
    MSMR_det = mass**-0.25 * np.exp(temps/10) * (0.2/6.04)
    Cort_det = mass**-0.25 * np.exp(temps/10) * (7.8/6.04)
    
    MSMR = np.random.gamma((MSMR_det**2)/(MSMRSD**2), (MSMRSD**2)/MSMR_det)
    Cort = np.random.gamma((Cort_det**2)/(CortSD**2), (CortSD**2)/Cort_det)

    df = pd.DataFrame({
        'ID': ids,
        'Temp': temps,
        'sex': sex,
        'Mass': mass,
        'MSMR': MSMR,
        'Cort': Cort
    })

    msmr_model = ols("np.log(MSMR) ~ np.log(Mass) + Temp", data=df).fit()
    cort_model = ols("np.log(Cort) ~ np.log(Mass) + Temp", data=df).fit()
    cort_msmr_model = ols("np.log(Cort) ~ np.log(MSMR)", data=df).fit()

    MSMR_models.append(msmr_model)
    Cort_models.append(cort_model)
    Cort_MSMR_models.append(cort_msmr_model)

# Save the last simulation
df.to_csv("SimulatedData.csv", index=False)

# Stats Summary
def average_coef_stats(models, predictors):
    means = {p: [] for p in predictors}
    rsq = []

    for model in models:
        rsq.append(model.rsquared)
        for p in predictors:
            means[p].append(model.params[p])
    
    means_avg = {p: np.mean(means[p]) for p in predictors}
    se = {p: np.std(means[p], ddof=1) for p in predictors}
    return means_avg, se, np.mean(rsq)

# Extract results
MSMR_avg, MSMR_se, MSMR_rsq = average_coef_stats(MSMR_models, ['Intercept', 'np.log(Mass)', 'Temp'])
Cort_avg, Cort_se, Cort_rsq = average_coef_stats(Cort_models, ['Intercept', 'np.log(Mass)', 'Temp'])
CortMSMR_avg, CortMSMR_se, CortMSMR_rsq = average_coef_stats(Cort_MSMR_models, ['Intercept', 'np.log(MSMR)'])

# Build stat table
stat_table = pd.DataFrame([
    [MSMR_avg['np.log(Mass)'], MSMR_se['np.log(Mass)'], '', '', MSMR_avg['Intercept'], MSMR_rsq],
    [MSMR_avg['Temp'], MSMR_se['Temp'], '', '', '', ''],
    [Cort_avg['np.log(Mass)'], Cort_se['np.log(Mass)'], '', '', Cort_avg['Intercept'], Cort_rsq],
    [Cort_avg['Temp'], Cort_se['Temp'], '', '', '', ''],
    [CortMSMR_avg['np.log(MSMR)'], CortMSMR_se['np.log(MSMR)'], '', '', CortMSMR_avg['Intercept'], CortMSMR_rsq],
], columns=["Estimate", "SE Est.", "T value", "p value", "Intercept", "R squared"], 
   index=["MSMR ~ mass", "MSMR ~ Temp", "Cort ~ mass", "Cort ~ Temp", "Cort ~ MSMR"])

# Round and format
stat_table = stat_table.round(3)
stat_table.loc[["MSMR ~ Temp", "Cort ~ Temp"], "Intercept"] = ""
stat_table.loc[["MSMR ~ Temp", "Cort ~ Temp"], "R squared"] = ""

# Save to CSV
os.makedirs("Power_Analysis", exist_ok=True)
stat_table.to_csv("StatsTab.csv")

# Plotting (Summary overlays)
def plot_model_lines(x, y, models, x_label, y_label, x_log=False, y_log=False, ylims=None):
    plt.figure(figsize=(5, 4))
    x_vals = np.linspace(min(x), max(x), 100)
    for model in models:
        intercept, slope = model.params[:2]
        y_vals = intercept + slope * np.log(x_vals) if x_log else intercept + slope * x_vals
        plt.plot(np.log(x_vals) if x_log else x_vals, y_vals, color='blue', alpha=0.02)
    plt.xlabel(x_label)
    plt.ylabel(y_label)
    if ylims:
        plt.ylim(ylims)
    plt.grid(False)
    plt.tight_layout()

plot_model_lines(df["Mass"], df["MSMR"], MSMR_models, "ln Body Mass (g)", "ln MSMR", x_log=True, y_log=True, ylims=(-7, 0))
plot_model_lines(df["Temp"], df["MSMR"], MSMR_models, "Temperature (C)", "ln MSMR", y_log=True)

plot_model_lines(df["Mass"], df["Cort"], Cort_models, "ln Body Mass (g)", "ln Cort", x_log=True, y_log=True, ylims=(-5, 2))
plot_model_lines(df["Temp"], df["Cort"], Cort_models, "Temperature (C)", "ln Cort", y_log=True)

plot_model_lines(df["MSMR"], df["Cort"], Cort_MSMR_models, "ln MSMR", "ln Cort", x_log=True, y_log=True)

plt.show()
